// zalo-bot-js/dist/src/request/BaseRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseRequest = void 0;
const errors_1 = require("../errors");
class BaseRequest {
    async post(url, data, options) {
        const response = await this.requestWrapper(url, "POST", data, options);
        const json = this.parseJsonPayload(response.body);
        return json.result;
    }
    async requestWrapper(url, method, data, options) {
        const retryPolicy = resolveRetryPolicy(options?.retryPolicy);
        let attempt = 1;
        while (attempt <= retryPolicy.maxAttempts) {
            const start = Date.now();
            const attemptContext = {
                url,
                method,
                data: data,
                attempt,
            };
            await options?.hooks?.beforeRequest?.(attemptContext);
            let response;
            let error;
            try {
                response = await this.doRequest(url, method, data, options);
                if (response.status >= 200 && response.status <= 299) {
                    await this.callAfterResponseHook(options, {
                        ...attemptContext,
                        durationMs: Date.now() - start,
                        responseStatus: response.status,
                        responseBody: response.body,
                    });
                    return response;
                }
                error = this.mapErrorFromResponse(response);
            }
            catch (caught) {
                error = this.normalizeRequestError(caught);
            }
            await this.callAfterResponseHook(options, {
                ...attemptContext,
                durationMs: Date.now() - start,
                responseStatus: response?.status,
                responseBody: response?.body,
                error,
            });
            if (!this.shouldRetry(error, retryPolicy, options?.signal) || attempt >= retryPolicy.maxAttempts) {
                throw error;
            }
            await sleep(this.computeRetryDelayMs(attempt, retryPolicy), options?.signal);
            attempt += 1;
        }
        throw new errors_1.NetworkError("Request retry loop exhausted unexpectedly");
    }
    parseJsonPayload(payload) {
        try {
            return JSON.parse(payload);
        }
        catch (error) {
            throw new errors_1.ZaloError(`Invalid server response: ${String(error)}`);
        }
    }
    mapErrorFromResponse(response) {
        const payload = this.parseJsonPayload(response.body);
        const description = String(payload.description ?? "Unknown HTTP error");
        const parameters = payload.parameters;
        if (parameters?.migrate_to_chat_id && typeof parameters.migrate_to_chat_id === "number") {
            return new errors_1.ChatMigrated(parameters.migrate_to_chat_id);
        }
        if (parameters?.retry_after && typeof parameters.retry_after === "number") {
            return new errors_1.RetryAfter(parameters.retry_after);
        }
        if (response.status === 403) {
            return new errors_1.Forbidden(description);
        }
        if (response.status === 401 || response.status === 404) {
            return new errors_1.InvalidToken(description);
        }
        if (response.status === 400) {
            return new errors_1.BadRequest(description);
        }
        if (response.status === 409) {
            return new errors_1.Conflict(description);
        }
        return new errors_1.NetworkError(`${description} (${response.status})`);
    }
    normalizeRequestError(error) {
        if (error instanceof errors_1.ZaloError) {
            return error;
        }
        return new errors_1.NetworkError(`Unknown error in HTTP implementation: ${String(error)}`);
    }
    shouldRetry(error, policy, signal) {
        if (signal?.aborted) {
            return false;
        }
        const kind = toRetryErrorKind(error);
        return policy.retryOn.includes(kind);
    }
    computeRetryDelayMs(attempt, policy) {
        const base = policy.baseDelayMs * 2 ** Math.max(attempt - 1, 0);
        const bounded = Math.min(policy.maxDelayMs, base);
        if (policy.jitterRatio <= 0) {
            return bounded;
        }
        const min = Math.max(0, bounded * (1 - policy.jitterRatio));
        const max = bounded * (1 + policy.jitterRatio);
        return Math.floor(min + Math.random() * (max - min));
    }
    async callAfterResponseHook(options, context) {
        await options?.hooks?.afterResponse?.(context);
    }
}
exports.BaseRequest = BaseRequest;
function resolveRetryPolicy(policy) {
    return {
        maxAttempts: Math.max(1, policy?.maxAttempts ?? 1),
        baseDelayMs: Math.max(0, policy?.baseDelayMs ?? 250),
        maxDelayMs: Math.max(0, policy?.maxDelayMs ?? 3000),
        jitterRatio: Math.max(0, Math.min(1, policy?.jitterRatio ?? 0.2)),
        retryOn: policy?.retryOn ?? ["timed_out", "network_error", "conflict", "retry_after"],
    };
}
function toRetryErrorKind(error) {
    if (error instanceof errors_1.TimedOut) {
        return "timed_out";
    }
    if (error instanceof errors_1.RetryAfter) {
        return "retry_after";
    }
    if (error instanceof errors_1.Conflict) {
        return "conflict";
    }
    if (error instanceof errors_1.BadRequest) {
        return "bad_request";
    }
    if (error instanceof errors_1.Forbidden) {
        return "forbidden";
    }
    if (error instanceof errors_1.InvalidToken) {
        return "invalid_token";
    }
    if (error instanceof errors_1.NetworkError) {
        return "network_error";
    }
    return "zalo_error";
}
async function sleep(ms, signal) {
    if (ms <= 0 || signal?.aborted) {
        return;
    }
    await new Promise((resolve) => {
        const timeout = setTimeout(() => {
            signal?.removeEventListener("abort", onAbort);
            resolve();
        }, ms);
        const onAbort = () => {
            clearTimeout(timeout);
            signal?.removeEventListener("abort", onAbort);
            resolve();
        };
        signal?.addEventListener("abort", onAbort, { once: true });
    });
}
