// zalo-bot-js/dist/src/request/FetchRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchRequest = void 0;
const errors_1 = require("../errors");
const BaseRequest_1 = require("./BaseRequest");
class FetchRequest extends BaseRequest_1.BaseRequest {
    config;
    readTimeout;
    constructor(config = {}) {
        super();
        this.config = config;
        this.readTimeout = config.readTimeout ?? 5000;
    }
    async initialize() { }
    async shutdown() { }
    async doRequest(url, method, data, options) {
        const timeoutMs = this.resolveTimeout(options);
        const timeoutController = new AbortController();
        const timeoutId = setTimeout(() => timeoutController.abort(), timeoutMs);
        const signal = mergeSignals(options?.signal, timeoutController.signal);
        try {
            const response = await fetch(url, {
                method,
                headers: {
                    "Content-Type": "application/json",
                    "User-Agent": "zalo-bot-js",
                },
                body: method === "POST" ? JSON.stringify(this.compactPayload(data)) : undefined,
                signal,
            });
            return {
                status: response.status,
                body: await response.text(),
            };
        }
        catch (error) {
            if (error instanceof Error && error.name === "AbortError") {
                if (options?.signal?.aborted) {
                    throw new errors_1.TimedOut("Request aborted");
                }
                throw new errors_1.TimedOut();
            }
            throw new errors_1.NetworkError(`fetch.${error instanceof Error ? error.message : String(error)}`);
        }
        finally {
            clearTimeout(timeoutId);
        }
    }
    resolveTimeout(options) {
        const explicitPolicy = options?.timeoutPolicy;
        const configPolicy = this.config.timeoutPolicy;
        const profile = options?.timeoutProfile ?? "standard";
        const profileTimeout = profile === "short"
            ? explicitPolicy?.shortMs ?? configPolicy?.shortMs
            : profile === "long_poll"
                ? explicitPolicy?.longPollMs ?? configPolicy?.longPollMs
                : explicitPolicy?.standardMs ?? configPolicy?.standardMs;
        if (typeof profileTimeout === "number" && Number.isFinite(profileTimeout)) {
            return Math.max(profileTimeout, 0);
        }
        const readTimeoutSeconds = options?.readTimeout;
        if (typeof readTimeoutSeconds === "number" && Number.isFinite(readTimeoutSeconds)) {
            return Math.max(readTimeoutSeconds, 0) * 1000;
        }
        return this.readTimeout ?? 5000;
    }
    compactPayload(data) {
        if (!data) {
            return {};
        }
        return Object.fromEntries(Object.entries(data).filter(([, value]) => value !== undefined && value !== null));
    }
}
exports.FetchRequest = FetchRequest;
function mergeSignals(signalA, signalB) {
    if (!signalA) {
        return signalB;
    }
    if (!signalB) {
        return signalA;
    }
    if (signalA.aborted || signalB.aborted) {
        const controller = new AbortController();
        controller.abort();
        return controller.signal;
    }
    const controller = new AbortController();
    const abort = () => {
        controller.abort();
        signalA.removeEventListener("abort", abort);
        signalB.removeEventListener("abort", abort);
    };
    signalA.addEventListener("abort", abort, { once: true });
    signalB.addEventListener("abort", abort, { once: true });
    return controller.signal;
}
