// zalo-bot-js/dist/src/models/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = void 0;
const Chat_1 = require("./Chat");
const User_1 = require("./User");
class Message {
    bot;
    messageId;
    date;
    chat;
    messageType;
    text;
    sticker;
    photoUrl;
    fromUser;
    admin;
    raw;
    constructor(init) {
        this.bot = init.bot;
        this.messageId = init.messageId;
        this.date = init.date;
        this.chat = init.chat;
        this.messageType = init.messageType ?? "CHAT_MESSAGE";
        this.text = init.text;
        this.sticker = init.sticker;
        this.photoUrl = init.photoUrl;
        this.fromUser = init.fromUser;
        this.admin = init.admin ?? false;
        this.raw = init.raw;
    }
    static fromApi(data, bot) {
        if (!data) {
            return undefined;
        }
        const messageId = typeof data.message_id === "string"
            ? data.message_id
            : typeof data.message_id === "number"
                ? String(data.message_id)
                : undefined;
        if (!messageId) {
            return undefined;
        }
        const chat = Chat_1.Chat.fromApi(asJsonObject(data.chat));
        if (!chat) {
            return undefined;
        }
        const timestamp = typeof data.date === "number" ? data.date : Date.now();
        const fromUser = User_1.User.fromApi(asJsonObject(data.from));
        return new Message({
            bot,
            messageId,
            date: new Date(timestamp),
            chat,
            messageType: typeof data.message_type === "string" ? data.message_type : undefined,
            text: typeof data.text === "string" ? data.text : undefined,
            sticker: typeof data.sticker === "string" ? data.sticker : undefined,
            photoUrl: typeof data.photo_url === "string" ? data.photo_url : undefined,
            fromUser,
            admin: bot?.isAdmin(fromUser?.id),
            raw: data,
        });
    }
    replyText(text) {
        return this.requireBot().sendMessage(this.chat.id, text);
    }
    replyPhoto(photo, caption = "") {
        return this.requireBot().sendPhoto(this.chat.id, caption, photo);
    }
    replySticker(sticker) {
        return this.requireBot().sendSticker(this.chat.id, sticker);
    }
    replyAction(action) {
        return this.requireBot().sendChatAction(this.chat.id, action);
    }
    requireBot() {
        if (!this.bot) {
            throw new Error("Message is not bound to a bot instance.");
        }
        return this.bot;
    }
}
exports.Message = Message;
function asJsonObject(value) {
    if (!value || typeof value !== "object" || Array.isArray(value)) {
        return undefined;
    }
    return value;
}
