// zalo-bot-js/dist/src/models/Update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Update = void 0;
exports.parseCommand = parseCommand;
const Message_1 = require("./Message");
class Update {
    updateId;
    message;
    raw;
    constructor(updateId, message, raw) {
        this.updateId = updateId;
        this.message = message;
        this.raw = raw;
    }
    get effectiveUser() {
        return this.message?.fromUser;
    }
    get eventTypes() {
        const eventTypes = new Set();
        if (this.message) {
            eventTypes.add("message");
        }
        if (this.message?.text) {
            eventTypes.add("text");
            if (this.command) {
                eventTypes.add("command");
            }
        }
        if (this.message?.photoUrl) {
            eventTypes.add("photo");
        }
        if (this.message?.sticker) {
            eventTypes.add("sticker");
        }
        return [...eventTypes];
    }
    hasEventType(eventType) {
        return this.eventTypes.includes(eventType);
    }
    get command() {
        return parseCommand(this.message?.text);
    }
    static fromApi(data, bot) {
        if (!data) {
            return undefined;
        }
        const updateId = typeof data.update_id === "number"
            ? data.update_id
            : typeof data.update_id === "string"
                ? Number.parseInt(data.update_id, 10)
                : undefined;
        return new Update(updateId, Message_1.Message.fromApi(asJsonObject(data.message), bot), data);
    }
}
exports.Update = Update;
function asJsonObject(value) {
    if (!value || typeof value !== "object" || Array.isArray(value)) {
        return undefined;
    }
    return value;
}
function parseCommand(text) {
    if (!text) {
        return undefined;
    }
    const normalized = text.trim();
    if (!normalized.startsWith("/") || normalized.length <= 1) {
        return undefined;
    }
    const body = normalized.slice(1);
    const [rawName, ...rest] = body.split(/\s+/);
    const name = rawName.trim();
    if (!name) {
        return undefined;
    }
    const argsRaw = rest.join(" ").trim();
    return {
        name,
        argsRaw,
        args: argsRaw ? argsRaw.split(/\s+/) : [],
    };
}
