// zalo-bot-js/dist/src/models/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
class User {
    id;
    displayName;
    accountName;
    accountType;
    isBot;
    canJoinGroups;
    raw;
    constructor(id, displayName, accountName, accountType, isBot, canJoinGroups, raw) {
        this.id = id;
        this.displayName = displayName;
        this.accountName = accountName;
        this.accountType = accountType;
        this.isBot = isBot;
        this.canJoinGroups = canJoinGroups;
        this.raw = raw;
    }
    static fromApi(data) {
        if (!data || typeof data.id !== "string") {
            return undefined;
        }
        return new User(data.id, typeof data.display_name === "string" ? data.display_name : undefined, typeof data.account_name === "string" ? data.account_name : undefined, typeof data.account_type === "string" ? data.account_type : undefined, typeof data.is_bot === "boolean" ? data.is_bot : undefined, typeof data.can_join_groups === "boolean" ? data.can_join_groups : undefined, data);
    }
}
exports.User = User;
