// zalo-bot-js/dist/src/core/Bot.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bot = void 0;
const constants_1 = require("../constants");
const errors_1 = require("../errors");
const runtime_1 = require("../i18n/runtime");
const Chat_1 = require("../models/Chat");
const Message_1 = require("../models/Message");
const Update_1 = require("../models/Update");
const User_1 = require("../models/User");
const WebhookInfo_1 = require("../models/WebhookInfo");
const FetchRequest_1 = require("../request/FetchRequest");
const promises_1 = __importDefault(require("node:fs/promises"));
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const sqlite3_1 = require("sqlite3");
class Bot {
    baseUrl;
    request;
    initialized = false;
    botUser;
    eventListeners = new Map();
    textListeners = [];
    commandListeners = [];
    errorHandlers = [];
    logger;
    pollingState = "idle";
    pollingTask;
    pollingAbortController;
    nextUpdateOffset;
    adminId = process.env.ZALO_BOT_ADMIN_ID?.trim() ?? "";
    dataRootDir;
    config;
    constructor(tokenOrConfig, options = {}) {
        const config = typeof tokenOrConfig === "string"
            ? { ...options, token: tokenOrConfig }
            : tokenOrConfig;
        this.config = config;
        if (!config.token) {
            throw new errors_1.InvalidToken((0, runtime_1.t)("error.invalidTokenInput"));
        }
        const rootUrl = config.baseUrl ?? constants_1.BASE_URL;
        this.baseUrl = `${rootUrl}/bot${config.token}`;
        this.request = [
            config.pollingRequest ?? new FetchRequest_1.FetchRequest(),
            config.request ?? new FetchRequest_1.FetchRequest(),
        ];
        this.logger = config.logger ?? defaultLogger;
        if (config.onError) {
            this.errorHandlers.push(config.onError);
        }
        if (config.polling) {
            const pollingOptions = typeof config.polling === "object" ? config.polling : undefined;
            queueMicrotask(() => {
                void this.startPolling(pollingOptions).catch((error) => {
                    this.reportError(error, {
                        kind: "request_temporary",
                        source: "polling",
                    });
                });
            });
        }
    }
    async initialize() {
        if (this.initialized) {
            return;
        }
        await Promise.all(this.request.map((transport) => transport.initialize()));
        try {
            await this.getMe();
        }
        catch (error) {
            if (error instanceof errors_1.InvalidToken) {
                throw new errors_1.InvalidToken((0, runtime_1.t)("error.rejectedToken", { token: this.config.token }));
            }
            throw error;
        }
        this.initialized = true;
    }
    async shutdown() {
        if (!this.initialized) {
            return;
        }
        await Promise.all(this.request.map((transport) => transport.shutdown()));
        this.initialized = false;
    }
    async getMe(options) {
        const result = await this.post("getMe", {}, options);
        const user = User_1.User.fromApi(asJsonObject(result));
        if (!user) {
            throw new errors_1.InvalidToken((0, runtime_1.t)("error.invalidGetMePayload"));
        }
        this.botUser = user;
        return user;
    }
    async getUpdate(params = {}, options) {
        const updates = await this.getUpdates(params, options);
        return updates[0];
    }
    async getUpdates(params = {}, options) {
        const result = await this.post("getUpdates", {
            timeout: params.timeout,
            offset: params.offset,
            limit: params.limit,
            allowed_updates: params.allowedUpdates?.join(","),
        }, {
            ...options,
            // Long polling must wait slightly longer than the API timeout parameter.
            readTimeout: (params.timeout ?? 0) + 5,
        });
        const payloads = asJsonObjectArray(result);
        return payloads.map((payload) => Update_1.Update.fromApi(payload, this)).filter(isDefined);
    }
    async sendMessage(chatId, text, options) {
        return this.sendMessageLike("sendMessage", {
            chat_id: chatId,
            text,
            reply_to_message_id: options?.reply_to_message_id,
        }, options?.requestOptions);
    }
    async sendPhoto(chatId, caption, photo, options) {
        return this.sendMessageLike("sendPhoto", {
            chat_id: chatId,
            caption,
            photo,
            reply_to_message_id: options?.reply_to_message_id,
        }, options?.requestOptions);
    }
    async sendSticker(chatId, sticker, options) {
        return this.sendMessageLike("sendSticker", {
            chat_id: chatId,
            sticker,
            reply_to_message_id: options?.reply_to_message_id,
        }, options?.requestOptions);
    }
    async sendPhotos(chatId, photos, caption = "", options = {}) {
        const sanitizedPhotos = photos.map((photo) => photo.trim()).filter(Boolean);
        if (sanitizedPhotos.length === 0) {
            return [];
        }
        const separator = options.separator ?? " ";
        const replies = [];
        for (let index = 0; index < sanitizedPhotos.length; index += 1) {
            const photo = sanitizedPhotos[index];
            const label = `[${index + 1}/${sanitizedPhotos.length}]`;
            const indexedCaption = caption ? `${label}${separator}${caption}` : label;
            const message = await this.sendPhoto(chatId, indexedCaption, photo, {
                reply_to_message_id: options.reply_to_message_id,
                requestOptions: options.requestOptions,
            });
            replies.push(message);
        }
        return replies;
    }
    async sendChatAction(chatId, action, options) {
        const result = await this.post("sendChatAction", {
            chat_id: chatId,
            action,
        }, options);
        return Boolean(result);
    }
    async setWebhook(url, secretToken, options) {
        const result = await this.post("setWebhook", {
            url,
            secret_token: secretToken,
            drop_pending_updates: options?.dropPendingUpdates,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async deleteWebhook(options) {
        const result = await this.post("deleteWebhook", {
            drop_pending_updates: options?.dropPendingUpdates,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async getWebhookInfo(options) {
        const result = await this.post("getWebhookInfo", undefined, options);
        return WebhookInfo_1.WebhookInfo.fromApi(asJsonObject(result));
    }
    async editMessageText(chatId, messageId, text, options) {
        return this.sendMessageLike("editMessageText", {
            chat_id: chatId,
            message_id: messageId,
            text,
        }, options?.requestOptions);
    }
    async deleteMessage(chatId, messageId, options) {
        const result = await this.post("deleteMessage", {
            chat_id: chatId,
            message_id: messageId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async pinMessage(chatId, messageId, options) {
        const result = await this.post("pinMessage", {
            chat_id: chatId,
            message_id: messageId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async unpinMessage(chatId, messageId, options) {
        const result = await this.post("unpinMessage", {
            chat_id: chatId,
            message_id: messageId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async banChatMember(chatId, userId, options) {
        const result = await this.post("banChatMember", {
            chat_id: chatId,
            user_id: userId,
            reason: options?.reason,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async unbanChatMember(chatId, userId, options) {
        const result = await this.post("unbanChatMember", {
            chat_id: chatId,
            user_id: userId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async promoteChatAdmin(chatId, userId, options) {
        const result = await this.post("promoteChatAdmin", {
            chat_id: chatId,
            user_id: userId,
            role: options?.role,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async demoteChatAdmin(chatId, userId, options) {
        const result = await this.post("demoteChatAdmin", {
            chat_id: chatId,
            user_id: userId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async setChatKeyboard(chatId, keyboard, options) {
        const result = await this.post("setChatKeyboard", {
            chat_id: chatId,
            keyboard: JSON.stringify(keyboard),
        }, options?.requestOptions);
        return Boolean(result);
    }
    async deleteChatKeyboard(chatId, options) {
        const result = await this.post("deleteChatKeyboard", {
            chat_id: chatId,
        }, options?.requestOptions);
        return Boolean(result);
    }
    async uploadFile(fileUrl, options) {
        const result = await this.post("uploadFile", {
            file_url: fileUrl,
        }, options?.requestOptions);
        return toFileInfo(asJsonObject(result));
    }
    async getFileInfo(fileId, options) {
        const result = await this.post("getFileInfo", {
            file_id: fileId,
        }, options?.requestOptions);
        return toFileInfo(asJsonObject(result));
    }
    async getFileDownloadUrl(fileId, options) {
        const result = await this.post("getFileDownloadUrl", {
            file_id: fileId,
        }, options?.requestOptions);
        const payload = asJsonObject(result);
        const candidate = payload?.file_url;
        return typeof candidate === "string" ? candidate : undefined;
    }
    on(event, callback) {
        const listeners = this.eventListeners.get(event) ?? [];
        listeners.push({ callback, once: false });
        this.eventListeners.set(event, listeners);
        return this;
    }
    once(event, callback) {
        const listeners = this.eventListeners.get(event) ?? [];
        listeners.push({ callback, once: true });
        this.eventListeners.set(event, listeners);
        return this;
    }
    off(event, callback) {
        const listeners = this.eventListeners.get(event);
        if (!listeners) {
            return this;
        }
        this.eventListeners.set(event, listeners.filter((listener) => listener.callback !== callback));
        return this;
    }
    onText(pattern, callback) {
        this.textListeners.push({ pattern, callback, once: false });
        return this;
    }
    offText(pattern, callback) {
        const patternSource = pattern.toString();
        for (let index = this.textListeners.length - 1; index >= 0; index -= 1) {
            const listener = this.textListeners[index];
            if (listener.pattern.toString() === patternSource && listener.callback === callback) {
                this.textListeners.splice(index, 1);
            }
        }
        return this;
    }
    command(command, callback) {
        const normalized = normalizeCommand(command);
        if (!normalized) {
            throw new Error("Command name must not be empty");
        }
        this.commandListeners.push({
            command: normalized,
            callback,
        });
        return this;
    }
    onError(callback) {
        this.errorHandlers.push(callback);
        return this;
    }
    async processUpdate(update) {
        const normalizedUpdate = update instanceof Update_1.Update ? update : Update_1.Update.fromApi(update, this);
        if (!normalizedUpdate?.message) {
            if (!(update instanceof Update_1.Update)) {
                await this.reportError(new Error("Unable to parse incoming update payload"), {
                    kind: "payload_parse",
                    source: "event_dispatch",
                });
            }
            return;
        }
        if (typeof normalizedUpdate.updateId === "number") {
            this.nextUpdateOffset = normalizedUpdate.updateId + 1;
        }
        try {
            await this.tryEnsureUserDataDatabase(normalizedUpdate.message);
        }
        catch (error) {
            await this.reportError(error, {
                kind: "request_temporary",
                source: "event_dispatch",
                update: normalizedUpdate,
            });
        }
        await this.tryHandleSetAdmin(normalizedUpdate.message);
        await this.tryHandleWhoAmI(normalizedUpdate.message);
        const metadata = {
            update: normalizedUpdate,
            command: normalizedUpdate.command,
        };
        for (const eventType of normalizedUpdate.eventTypes) {
            const listeners = [...(this.eventListeners.get(eventType) ?? [])];
            for (const listener of listeners) {
                try {
                    await listener.callback(normalizedUpdate.message, metadata);
                }
                catch (error) {
                    await this.reportError(error, {
                        kind: "listener_user_code",
                        source: "event_dispatch",
                        update: normalizedUpdate,
                        event: eventType,
                        command: normalizedUpdate.command,
                    });
                }
                finally {
                    if (listener.once) {
                        this.off(eventType, listener.callback);
                    }
                }
            }
        }
        if (normalizedUpdate.message.text) {
            for (const listener of [...this.textListeners]) {
                const match = createRegexMatcher(listener.pattern).exec(normalizedUpdate.message.text);
                if (match) {
                    try {
                        await listener.callback(normalizedUpdate.message, match);
                    }
                    catch (error) {
                        await this.reportError(error, {
                            kind: "listener_user_code",
                            source: "text_listener",
                            update: normalizedUpdate,
                            command: normalizedUpdate.command,
                        });
                    }
                    finally {
                        if (listener.once) {
                            this.offText(listener.pattern, listener.callback);
                        }
                    }
                }
            }
        }
        if (normalizedUpdate.command) {
            for (const listener of this.commandListeners) {
                if (listener.command !== normalizedUpdate.command.name.toLowerCase()) {
                    continue;
                }
                try {
                    await listener.callback(normalizedUpdate.message, {
                        command: normalizedUpdate.command,
                        update: normalizedUpdate,
                    });
                }
                catch (error) {
                    await this.reportError(error, {
                        kind: "listener_user_code",
                        source: "command_listener",
                        update: normalizedUpdate,
                        command: normalizedUpdate.command,
                        event: "command",
                    });
                }
            }
        }
    }
    startPolling(options = {}) {
        if (this.pollingTask) {
            return this.pollingTask;
        }
        this.pollingState = "starting";
        this.pollingAbortController = new AbortController();
        this.pollingTask = this.runPolling(options).finally(() => {
            this.pollingState = "idle";
            this.pollingTask = undefined;
            this.pollingAbortController = undefined;
        });
        return this.pollingTask;
    }
    stopPolling() {
        if (this.pollingState === "idle") {
            return;
        }
        this.pollingState = "stopping";
        this.pollingAbortController?.abort();
    }
    isPolling() {
        return this.pollingState !== "idle";
    }
    getPollingState() {
        return this.pollingState;
    }
    async setWebHook(url, options) {
        return this.setWebhook(url, options?.secret_token ?? "", {
            dropPendingUpdates: options?.drop_pending_updates,
            requestOptions: options?.requestOptions,
        });
    }
    async deleteWebHook(options) {
        return this.deleteWebhook({
            dropPendingUpdates: options?.drop_pending_updates,
            requestOptions: options?.requestOptions,
        });
    }
    async getWebHookInfo(options) {
        return this.getWebhookInfo(options);
    }
    get cachedUser() {
        return this.botUser;
    }
    getAdminId() {
        return this.adminId || undefined;
    }
    isAdmin(userId) {
        if (!userId || !this.adminId) {
            return false;
        }
        return userId === this.adminId;
    }
    async reportError(error, context, overrideHandler) {
        this.logger.error((0, runtime_1.t)("app.pollingFetchError"), error, context);
        const handlers = [
            ...(overrideHandler ? [overrideHandler] : []),
            ...this.errorHandlers,
        ];
        for (const handler of handlers) {
            try {
                await handler(error, context);
            }
            catch (handlerError) {
                this.logger.error("Unhandled error from bot onError callback", handlerError, context);
            }
        }
    }
    async runPolling(options) {
        const timeoutSeconds = options.timeoutSeconds ?? constants_1.DEFAULT_POLL_TIMEOUT_SECONDS;
        const retryDelayMs = options.retryDelayMs ?? constants_1.DEFAULT_RETRY_DELAY_MS;
        const errorHandler = options.onError;
        await this.initialize();
        if (this.pollingState === "stopping") {
            return;
        }
        this.pollingState = "running";
        await this.logStartupStatus(options.requestOptions);
        try {
            while (this.pollingState === "running") {
                try {
                    const updates = await this.getUpdates({
                        timeout: timeoutSeconds,
                        offset: this.nextUpdateOffset,
                        allowedUpdates: options.allowedUpdates,
                    }, {
                        ...options.requestOptions,
                        signal: this.pollingAbortController?.signal,
                        timeoutProfile: options.requestOptions?.timeoutProfile ?? "long_poll",
                    });
                    if (updates.length > 0) {
                        for (const update of updates) {
                            if (options.onUpdate) {
                                try {
                                    await options.onUpdate(update);
                                }
                                catch (error) {
                                    await this.reportError(error, {
                                        kind: "listener_user_code",
                                        source: "polling",
                                        update,
                                    }, errorHandler);
                                }
                            }
                            await this.processUpdate(update);
                            if (this.pollingState !== "running") {
                                break;
                            }
                        }
                        continue;
                    }
                }
                catch (error) {
                    if (!(error instanceof errors_1.TimedOut) && this.pollingState === "running") {
                        await this.reportError(error, {
                            kind: "request_temporary",
                            source: "polling",
                        }, errorHandler);
                    }
                }
                if (this.pollingState !== "running") {
                    break;
                }
                await sleepWithAbort(retryDelayMs, this.pollingAbortController?.signal);
            }
        }
        finally {
            await this.shutdown();
        }
    }
    async sendMessageLike(endpoint, data, options) {
        const result = await this.post(endpoint, data, options);
        const rawResult = asJsonObject(result);
        const message = Message_1.Message.fromApi(rawResult, this);
        if (!message) {
            return this.buildMessageFallback(data, rawResult);
        }
        return message;
    }
    buildMessageFallback(data, result) {
        const chatId = typeof data.chat_id === "string" ? data.chat_id : undefined;
        if (!chatId) {
            throw new Error((0, runtime_1.t)("error.invalidMessageWithoutChatContext"));
        }
        const messageIdValue = result?.message_id;
        const timestampValue = result?.date;
        return new Message_1.Message({
            bot: this,
            messageId: typeof messageIdValue === "string"
                ? messageIdValue
                : typeof messageIdValue === "number"
                    ? String(messageIdValue)
                    : `local-${Date.now()}`,
            date: typeof timestampValue === "number"
                ? new Date(timestampValue)
                : new Date(),
            chat: new Chat_1.Chat(chatId, "direct"),
            messageType: "CHAT_MESSAGE",
            text: typeof data.text === "string" ? data.text : undefined,
            sticker: typeof data.sticker === "string" ? data.sticker : undefined,
            photoUrl: typeof data.photo === "string" ? data.photo : undefined,
            raw: result,
        });
    }
    async post(endpoint, data, options) {
        const request = endpoint === "getUpdates" ? this.request[0] : this.request[1];
        return request.post(`${this.baseUrl}/${endpoint}`, compactPayload(data), options);
    }
    async tryHandleSetAdmin(message) {
        const text = message.text?.trim().toLowerCase();
        if (text !== "/setadmin") {
            return;
        }
        if (this.adminId) {
            await this.sendMessage(message.chat.id, "Admin Bot da duoc set truoc do, khong the set lai.");
            return;
        }
        const candidateAdminId = message.fromUser?.id;
        if (!candidateAdminId) {
            await this.sendMessage(message.chat.id, "Khong lay duoc ID account. Thu lai voi tai khoan ca nhan.");
            return;
        }
        await writeOrUpdateEnv("ZALO_BOT_ADMIN_ID", candidateAdminId);
        this.adminId = candidateAdminId;
        process.env.ZALO_BOT_ADMIN_ID = candidateAdminId;
        await this.sendMessage(message.chat.id, `Admin Bot da duoc set: ${candidateAdminId}`);
    }
    async tryEnsureUserDataDatabase(message) {
        const userId = message.fromUser?.id;
        if (!userId) {
            return;
        }
        const rootDir = await this.resolveDataRootDir();
        const databaseFile = node_path_1.default.join(rootDir, `${sanitizeFileName(userId)}.db`);
        await ensureSqliteDatabase(databaseFile, userId);
    }
    async resolveDataRootDir() {
        if (this.dataRootDir) {
            return this.dataRootDir;
        }
        const me = this.botUser ?? (await this.getMe());
        const botLabel = me.displayName ?? me.accountName ?? me.id;
        const dirName = `Data${sanitizeFileName(botLabel)}`;
        const rootDir = node_path_1.default.resolve(process.cwd(), dirName);
        await promises_1.default.mkdir(rootDir, { recursive: true });
        this.dataRootDir = rootDir;
        return rootDir;
    }
    async tryHandleWhoAmI(message) {
        const text = message.text?.trim().toLowerCase();
        if (text !== "/id") {
            return;
        }
        const accountId = message.fromUser?.id;
        if (!accountId) {
            await this.sendMessage(message.chat.id, "Khong lay duoc ID account.");
            return;
        }
        await this.sendMessage(message.chat.id, `ID cua ban: ${accountId}\nadmin=${this.isAdmin(accountId)}`);
    }
    async logStartupStatus(options) {
        const pingStart = Date.now();
        const user = await this.getMe(options);
        const pingMs = Date.now() - pingStart;
        const ramMb = process.memoryUsage().rss / (1024 * 1024);
        const serverIp = getServerIpAddress();
        const country = await getCountryByIp(serverIp);
        const packageInfo = await getPackageInfo();
        const versionStatus = await getVersionStatus(packageInfo.name, packageInfo.version);
        const botName = user.displayName ?? user.accountName ?? "Unknown";
        const adminLabel = this.adminId ? `Da set (${this.adminId})` : "Chua set";
        console.log("[Zalo-bot-js] >> Bot khoi dong thanh cong...");
        console.log(`[Zalo-bot-js] >> Phien ban dang chay: ${versionStatus}`);
        console.log(`[Zalo-bot-js] >> Ban dang chay tren server: ${country}`);
        console.log(`[Zalo-bot-js] >> Bot Name: ${botName} (${user.id})`);
        console.log(`[Zalo-bot-js] >> Admin Bot: ${adminLabel}`);
        console.log(`[Zalo-bot-js] >> Ping: ${pingMs}ms`);
        console.log(`[Zalo-bot-js] >> Ram: ${ramMb.toFixed(2)}MB`);
    }
}
exports.Bot = Bot;
function compactPayload(data) {
    if (!data) {
        return {};
    }
    return Object.fromEntries(Object.entries(data).filter(([, value]) => value !== undefined && value !== null));
}
function asJsonObject(value) {
    if (!value || typeof value !== "object" || Array.isArray(value)) {
        return undefined;
    }
    return value;
}
function asJsonObjectArray(value) {
    if (Array.isArray(value)) {
        return value.filter((item) => Boolean(item) && typeof item === "object");
    }
    const singleValue = asJsonObject(value);
    return singleValue ? [singleValue] : [];
}
function isDefined(value) {
    return value !== undefined;
}
function createRegexMatcher(pattern) {
    const flags = pattern.flags.includes("g") ? pattern.flags : `${pattern.flags}g`;
    return new RegExp(pattern.source, flags);
}
function normalizeCommand(command) {
    const normalized = command.trim().replace(/^\//, "");
    return normalized.toLowerCase();
}
function getServerIpAddress() {
    const interfaces = node_os_1.default.networkInterfaces();
    for (const entries of Object.values(interfaces)) {
        if (!entries) {
            continue;
        }
        for (const info of entries) {
            if (info.family === "IPv4" && !info.internal) {
                return info.address;
            }
        }
    }
    return "127.0.0.1";
}
async function getCountryByIp(ip) {
    try {
        const response = await fetch(`https://api.country.is/${ip}`);
        if (!response.ok) {
            return "unknown";
        }
        const payload = (await response.json());
        return typeof payload.country === "string" ? payload.country : "unknown";
    }
    catch {
        return "unknown";
    }
}
let cachedPackageVersion;
let cachedPackageName;
async function getPackageInfo() {
    if (cachedPackageVersion && cachedPackageName) {
        return { name: cachedPackageName, version: cachedPackageVersion };
    }
    try {
        const packagePath = node_path_1.default.resolve(process.cwd(), "package.json");
        const content = await promises_1.default.readFile(packagePath, "utf8");
        const parsed = JSON.parse(content);
        cachedPackageName = typeof parsed.name === "string" ? parsed.name : "zalo-bot-js";
        cachedPackageVersion = typeof parsed.version === "string" ? parsed.version : "unknown";
        return { name: cachedPackageName, version: cachedPackageVersion };
    }
    catch {
        return { name: "zalo-bot-js", version: "unknown" };
    }
}
async function getVersionStatus(packageName, currentVersion) {
    try {
        const response = await fetch(`https://registry.npmjs.org/${encodeURIComponent(packageName)}/latest`);
        if (!response.ok) {
            return "Khong kiem tra duoc ban moi (npm registry loi).";
        }
        const payload = (await response.json());
        const latestVersion = typeof payload.version === "string" ? payload.version : undefined;
        if (!latestVersion || currentVersion === "unknown") {
            return "Khong xac dinh duoc phien ban npm moi nhat.";
        }
        const compareResult = compareSemver(currentVersion, latestVersion);
        if (compareResult >= 0) {
            return `Moi nhat ${currentVersion}`;
        }
        return `Co ban moi ${latestVersion}. Hien tai ${currentVersion}. Chay: npm i ${packageName}@latest`;
    }
    catch {
        return "Khong kiem tra duoc ban moi (mat ket noi).";
    }
}
function compareSemver(a, b) {
    const aCore = a.split("-")[0];
    const bCore = b.split("-")[0];
    const aParts = aCore.split(".").map((part) => Number.parseInt(part, 10) || 0);
    const bParts = bCore.split(".").map((part) => Number.parseInt(part, 10) || 0);
    const length = Math.max(aParts.length, bParts.length);
    for (let index = 0; index < length; index += 1) {
        const aValue = aParts[index] ?? 0;
        const bValue = bParts[index] ?? 0;
        if (aValue > bValue) {
            return 1;
        }
        if (aValue < bValue) {
            return -1;
        }
    }
    return 0;
}
async function writeOrUpdateEnv(key, value) {
    const envPath = node_path_1.default.resolve(process.cwd(), ".env");
    let content = "";
    try {
        content = await promises_1.default.readFile(envPath, "utf8");
    }
    catch {
        content = "";
    }
    const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const matcher = new RegExp(`^${escapedKey}=.*$`, "m");
    const nextLine = `${key}=${value}`;
    if (matcher.test(content)) {
        content = content.replace(matcher, nextLine);
    }
    else {
        const suffix = content.endsWith("\n") || content.length === 0 ? "" : "\n";
        content = `${content}${suffix}${nextLine}\n`;
    }
    await promises_1.default.writeFile(envPath, content, "utf8");
}
function sanitizeFileName(value) {
    const trimmed = value.trim();
    if (!trimmed) {
        return "unknown";
    }
    return trimmed.replace(/[^a-zA-Z0-9._-]/g, "_");
}
async function ensureSqliteDatabase(databasePath, userId) {
    await new Promise((resolve, reject) => {
        const database = new sqlite3_1.Database(databasePath, (openError) => {
            if (openError) {
                reject(openError);
                return;
            }
            database.serialize(() => {
                database.run("CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT NOT NULL)", (tableError) => {
                    if (tableError) {
                        database.close(() => reject(tableError));
                        return;
                    }
                    database.run("INSERT OR IGNORE INTO meta (key, value) VALUES ('user_id', ?)", [userId], (insertError) => {
                        if (insertError) {
                            database.close(() => reject(insertError));
                            return;
                        }
                        database.close((closeError) => {
                            if (closeError) {
                                reject(closeError);
                                return;
                            }
                            resolve();
                        });
                    });
                });
            });
        });
    });
}
const defaultLogger = {
    error(message, error) {
        console.error(message, error);
    },
};
async function sleepWithAbort(ms, signal) {
    if (!signal) {
        await sleep(ms);
        return;
    }
    if (signal.aborted) {
        return;
    }
    await new Promise((resolve) => {
        const timeout = setTimeout(() => {
            signal.removeEventListener("abort", onAbort);
            resolve();
        }, ms);
        const onAbort = () => {
            clearTimeout(timeout);
            signal.removeEventListener("abort", onAbort);
            resolve();
        };
        signal.addEventListener("abort", onAbort, { once: true });
    });
}
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
function toFileInfo(value) {
    if (!value) {
        return undefined;
    }
    const fileId = value.file_id;
    if (typeof fileId !== "string") {
        return undefined;
    }
    return {
        fileId,
        filePath: typeof value.file_path === "string" ? value.file_path : undefined,
        fileSize: typeof value.file_size === "number" ? value.file_size : undefined,
        raw: value,
    };
}
