// zalo-bot-js/dist/src/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Conflict = exports.RetryAfter = exports.ChatMigrated = exports.TimedOut = exports.BadRequest = exports.NetworkError = exports.InvalidToken = exports.Forbidden = exports.ZaloError = void 0;
class ZaloError extends Error {
    constructor(message) {
        super(ZaloError.normalizeMessage(message));
        this.name = new.target.name;
    }
    static normalizeMessage(message) {
        return message
            .replace(/^Error:\s*/i, "")
            .replace(/^\[Error\]:\s*/i, "")
            .replace(/^Bad Request:\s*/i, "");
    }
}
exports.ZaloError = ZaloError;
class Forbidden extends ZaloError {
}
exports.Forbidden = Forbidden;
class InvalidToken extends ZaloError {
    constructor(message = "Invalid token") {
        super(message);
    }
}
exports.InvalidToken = InvalidToken;
class NetworkError extends ZaloError {
}
exports.NetworkError = NetworkError;
class BadRequest extends NetworkError {
}
exports.BadRequest = BadRequest;
class TimedOut extends NetworkError {
    constructor(message = "Timed out") {
        super(message);
    }
}
exports.TimedOut = TimedOut;
class ChatMigrated extends ZaloError {
    newChatId;
    constructor(newChatId) {
        super(`Group migrated to supergroup. New chat id: ${newChatId}`);
        this.newChatId = newChatId;
    }
}
exports.ChatMigrated = ChatMigrated;
class RetryAfter extends ZaloError {
    retryAfter;
    constructor(retryAfter) {
        super(`Flood control exceeded. Retry in ${retryAfter} seconds`);
        this.retryAfter = retryAfter;
    }
}
exports.RetryAfter = RetryAfter;
class Conflict extends ZaloError {
}
exports.Conflict = Conflict;
